### DISCUSSION TITLE
v5-beta

### DISCUSSION THREAD

**Discussion by megahertz** [megahertz/electron-log#333] [General]
Hey guys. There are a lot of changes in v5 and I would appreciate any feedback. The most important things is a renderer-side part. I tried many approach how to make a renderer part pretty convenient, but maybe you have a better idea.

**Comment by suchcodemuchwow** [comment]
I have v5 installed but it gives me an error on renderer whenever I try import any of the following:
- `import log from 'electron-log';`
- `import log from 'electron-log/renderer';`
- `import log from 'electron-log/src/renderer';`

I'm using **electron-log** in this boilerplate [electron-react-boilerplate](https://github.com/electron-react-boilerplate)

Error on renderer:
```
Module not found: Can't resolve 'fs' in '/Users/batu/Desktop/spwn/node_modules/electron'
```

Error on main:
```
ERROR in ./node_modules/electron/index.js 1:11-24
Module not found: Error: Can't resolve 'fs' in '/Users/batu/Desktop/spwn/node_modules/electron'
 @ ./node_modules/electron-log/src/renderer.js 21:36-55
 @ ./src/renderer/Login.tsx 6:0-31 9:8-16
 @ ./src/renderer/App.tsx 7:0-28 18:230-235
 @ ./src/renderer/index.tsx 6:0-24 9:17-20

ERROR in ./node_modules/electron/index.js 2:13-28
Module not found: Error: Can't resolve 'path' in '/Users/batu/Desktop/spwn/node_modules/electron'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "path": require.resolve("path-browserify") }'
	- install 'path-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "path": false }
 @ ./node_modules/electron-log/src/renderer.js 21:36-55
 @ ./src/renderer/Login.tsx 6:0-31 9:8-16
 @ ./src/renderer/App.tsx 7:0-28 18:230-235
 @ ./src/renderer/index.tsx 6:0-24 9:17-20

webpack compiled with 2 errors
```

**Reply by megahertz** [reply]
That's strange. Have you changed the default webpack config? It should work fine when target is set to 'electron-renderer'

**Reply by suchcodemuchwow** [reply]
No 😔, I haven't touched webpack config which exactly looks like the same in boilerplate [webpack.config.renderer.dev.ts](https://github.com/electron-react-boilerplate/electron-react-boilerplate/blob/main/.erb/configs/webpack.config.renderer.dev.ts#L46)

**Reply by megahertz** [reply]
Thank you. Will try to find a solution. https://github.com/megahertz/electron-log/issues/335

**Reply by megahertz** [reply]
fixed

**Comment by maximevince** [comment]
Does this work with `contextIsolation` enabled? In that case one cannot import / require in the renderer.js.
How to enable the logging in the renderer process, in that case? Or in that case, from the renderer, send the logging calls to the main through IPC?

**Reply by megahertz** [reply]
Yes it works https://github.com/megahertz/electron-log/blob/master/docs/initialize.md

**Comment by fool-derek** [comment]
When will version 5.0.0 be released?

**Reply by megahertz** [reply]
May be in February. Not sure currently.

**Comment by jupnit** [comment]
Hi, I am using a bundler for both preload and main files. Hence, instead of initializing using - 
`log.initialize({ preload: true })` , 
I am importing logger in preload as suggested in docs - 
`import log from 'electron-log/renderer'`

Everything is working fine for me but I am still getting this error which should be suppressed as I have intentionally not initialized the logger for renderer - 

`electron-log: logger isn't initialized in the main process`

**Reply by megahertz** [reply]
It depends on your contextIsolation/sandbox settings. I'm going to improve docs later, currently I'm not sure how to describe it better.

Anyway, in the main process you need to call `import 'electron-log'` at least.

**Comment by g-bastianelli** [comment]
Hello, 
I have this issue when i try to use electron-log
```
 [plugin-webpack] Launching dev servers for renderer process code
[electron.serve] [DATA] Output Available: http://localhost:10000
[electron.serve] [DATA]
[electron.serve] [SUCCESS] [plugin-webpack] Launching dev servers for renderer process code
[electron.serve] App threw an error during load
[electron.serve] Error: Module build failed (from ./node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js):
[electron.serve] SyntaxError: Unexpected token (66:30)
[electron.serve]     at module.exports.pp$4.raise (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:20834:13)
[electron.serve]     at module.exports.pp.unexpected (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:18680:8)
[electron.serve]     at module.exports.pp$3.parseExprAtom (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:20282:10)
[electron.serve]     at _class.parseExprAtom (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:4551:117)
[electron.serve]     at _class.parseExprAtom (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:42293:75)
[electron.serve]     at _class.parseExprAtom (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:35894:26)
[electron.serve]     at _class.parseExprAtom (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:15975:26)
[electron.serve]     at _class.parseExprAtom (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:380:26)
[electron.serve]     at module.exports.pp$3.parseExprSubscripts (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:20116:19)
[electron.serve]     at module.exports.pp$3.parseMaybeUnary (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:20093:17)
[electron.serve]     at _class.parseMaybeUnary (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:35875:29)
[electron.serve]     at _class.parseMaybeUnary (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:15955:29)
[electron.serve]     at _class.parseMaybeUnary (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:368:29)
[electron.serve]     at module.exports.pp$3.parseExprOps (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:20035:19)
[electron.serve]     at module.exports.pp$3.parseMaybeConditional (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:20018:19)
[electron.serve]     at module.exports.pp$3.parseMaybeAssign (/Users/gbastianelli/Projects/modjo-desktop/node_modules/@marshallofsound/webpack-asset-relocator-loader/dist/index.js:19993:19)
[electron.serve]     at ./node_modules/electron-log/src/main/index.js (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:31760:7)
[electron.serve]     at __webpack_require__ (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:93412:42)
[electron.serve]     at ./node_modules/electron-log/src/index.js (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:31748:20)
[electron.serve]     at __webpack_require__ (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:93412:42)
[electron.serve]     at ./apps/electron/src/main/logger.ts (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:63410:70)
[electron.serve]     at __webpack_require__ (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:93412:42)
[electron.serve]     at ./apps/electron/src/main/NoteManager/index.ts (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:61864:66)
[electron.serve]     at __webpack_require__ (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:93412:42)
[electron.serve]     at ./apps/electron/src/main/index.ts (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:63277:71)
[electron.serve]     at __webpack_require__ (/Users/gbastianelli/Projects/modjo-desktop/.webpack/main/index.js:93412:42)
```
i m not using the log in the renderer process. I work with elecron-forge.

**Reply by megahertz** [reply]
How can I reproduce the issue?

**Reply by g-bastianelli** [reply]
It's ok, i just add to change a dependencies @marshallofsound/webpack-asset-relocator-loader to @vercel/webpack-asset-relocator-loader. And everything works perfectly now. Sorry for disturbance

**Reply by dj-fiorex** [reply]
Sorry to reopen this but we need to use this fork (@marshallofsound/webpack-asset-relocator-loader)  of the vercel  webpack-asset-relocator-loader because with it we can load native modules through electron-forge starter template. Do you have some clue on why this is happening? 
To reproduce just create a new electron-forge typescript + webpack app using the template, than substitute `@vercel/webpack-asset-relocator-loader` with `@marshallofsound/webpack-asset-relocator-loader`

Thanks!

**Reply by megahertz** [reply]
@dj-fiorex Added an issue https://github.com/megahertz/electron-log/issues/353

**Reply by megahertz** [reply]
Just updated the issue

**Comment by itsdouges** [comment]
Hey mate trying it out on my new Electron app. Seems to work well, using:

- startCatching()
- initialize({ preload: true, spyRendererConsole: true })
- Object.assign(console, log.functions)

All log as expected. What isn't expected is I'm not sure where the logs are persisting to on the fs (I don't see anything in `~/.config`). Does it log in dev or only when bundled? Is there any need to set `log.transports.file.level = "info"` or that's redundant? (Following https://github.com/iffy/electron-updater-example for context, might be outdated)

**Reply by megahertz** [reply]
You may try calling `log.transports.file.getFile().path` in the main process to get the actual path.

**Reply by itsdouges** [reply]
That throws. Perhaps the file path isn't set by default accidentally?

```
16:57:26.963 › Unhandled TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
```

**Reply by megahertz** [reply]
That's strange. Could you please create a new issue and provide more details here?

**Reply by itsdouges** [reply]
Not sure why it's throwing, doing some splunking though I found the path it logs to - `/Users/douges/Library/Logs/Triplex` 😄 

Edit: I now realize I was looking at Linux log paths 🤦 sorry folks. However it does look to be logging inside my user directory not root ~/Library/Logs/{app name}/main.log if that makes a difference. Is it conditional?

**Comment by tripower** [comment]
Hi

I have installed 
v5.0.0-beta.2

and in **main.js**
**const log = require('electron-log');
log.initialize({ spyRendererConsole: true});** 

but now I get following error and app doesn't start :(
**App threw an error during load
Error: Either preload or spyRendererConsole should be set
    at initialize (C:\nodejs\electron-quick-start\node_modules\electron-log\src\core\initialize.js:13:13)
    at Logger.initialize (C:\nodejs\electron-quick-start\node_modules\electron-log\src\core\Logger.js:100:5)**

any ideas whats wrong?

**Reply by megahertz** [reply]
Hi, you need to install the latest beta

**Reply by tripower** [reply]
thx 
with **v5.0.0-beta.23**
the **app starts**

should now all what I see in the console of the DevTools be written into a file?
because I see only my test call from the main.js

**Reply by megahertz** [reply]
Yes, if it doesn't work for you please create a new issue with a sample to reproduce it

**Reply by tripower** [reply]
I think that for my demand
the switch from chrome is now enough

and I have only the 2 lines in my code
hoped that this would be enough for your logging

**main.js**
const log = require('electron-log');
log.initialize({ spyRendererConsole: true});

**Reply by megahertz** [reply]
Just to link this comment to the related issue #361

**Comment by tripower** [comment]
another question
should a log entry from the renderer process not be defined so?
format => **'{h}:{i}:{s} {processType} {text}'**
Output:
**12:48:14 main Hello, log
12:48:24 main log from web by kb
12:48:24 main info from web by kb
12:48:24 main error from web by kb**

Update 
have now edited your code (local) in the **function initializeSpyRendererConsole** 
to 
**logger.processMessage({ data: ['renderer: ' + message], level: levels[level] });**
which would bring the estimated result
15:16:29 file: [main] [] Hello, log
15:16:39 file: [main] [] **renderer:** log from web by kb

I'm also not sure, if it is ok 
that all is in the file main.log??
should so not also be a renderer.log for the messages of SpyRenderer??

**Reply by megahertz** [reply]
You're right, thanks for the information. Fixed in the latest build.

**Reply by tripower** [reply]
thx
thats great 🤗

**Reply by tripower** [reply]
thx 
works 
09:50:50 file: [main] [] Hello, log
09:51:06 file: [renderer] [] log from web by kb

but should both be in one file?
or normally 
main.log and renderer.log?

**Reply by megahertz** [reply]
In v5 it's written to the same file

**Reply by tripower** [reply]
ok

would be a nice feature
to separate it or an option for main and renderer filename / path
only wondering because in the code the method getDefaultFileName returns for every processType a different filename
and the processTyp is now different (main, renderer)

but thanks for your quick other fix!

**Reply by megahertz** [reply]
It can be configured through [`resolvePathFn`](https://github.com/megahertz/electron-log/blob/master/docs/transports/file.md#resolvepathfn-variables-pathvariables-message-logmessage--string), just check message.variables.processType field to set the proper file path.

**Reply by tripower** [reply]
wow, thanks for the hint
that's perfect
**powerful tool!**

**Reply by tripower** [reply]
a common question
have you anytime made a performance test
if the logging do much slow down the apps?

**Reply by megahertz** [reply]
I made it for early versions of v5. Not a serious testing, just to make sure it have no critical bottleneck. It's fine for most cases. For the apps which produces more than 1mb logs per seconds it doesn't make sense to use such a library.

**Reply by tripower** [reply]
ok. good to know
but so much logs should never be written in my case
so seems to be ok
thx

**Reply by tripower** [reply]
@megahertz is there also a way to disable logging for a specific processType?
I found only in the doc to disable a transport complete by the level

**Reply by megahertz** [reply]
There's no straightforward way but there are some workaround depending on you requirements. For example, you can use a custom hook for that.

**Reply by tripower** [reply]
ok
will take a look
for hook 🧐

**Reply by tripower** [reply]
@megahertz 
Hook should work for me
but little bit confusing 

Code:
```
log.hooks.push((message, transport) => {
  if (transport == log.transports.file) {
     message.data[0] = "-file- " + message.data[0] + "- ";
  } else if (transport == log.transports.console) {
     message.data[0] = "-console- " + message.data[0]+ "- ";
  }

  if (message.variables.processType == 'renderer') {
     message.data[0] = "-renderer- " + message.data[0]+ "- ";
  } else {
     message.data[0] = "-main- " + message.data[0]+ "- ";
  }
  //console.log("hooks message:   " + JSON.stringify(message, null, 3));
  //console.log("hooks transport1: " + JSON.stringify(transport, null, 3));
  //console.log("hooks transport2: " + transport);

  return message;
});
```

**Output in console:** 
14:42:45 [renderer] -renderer- -console- log from web by kb- -
**Output in file renderer.log:**
14:42:45 [renderer] -renderer- -file- -renderer- -console- log from web by kb- - - -

my **expectation** was in the file
14:42:45 [renderer] -renderer- -file- log from web by kb- -

**Reply by megahertz** [reply]
You shouldn't modify data directly since this array is used for each transport. If you want to do that you need to clone this array instead and return a modified message.

**Reply by tripower** [reply]
@megahertz 
I have tried now this
but same problem
can you post an example please, thx

```
log.hooks.push((message, transport) => {
  let newMessage = Object.assign({}, message);
  if (transport == log.transports.file) {
    newMessage.data[0] = "-file- " + newMessage.data[0] + "- ";
  } else if (transport == log.transports.console) {
    newMessage.data[0] = "-console- " + newMessage.data[0] + "- ";
  }

  if (message.variables.processType == 'renderer') {
    newMessage.data[0] = "-renderer- " + newMessage.data[0] + "- ";
  } else {
    newMessage.data[0] = "-main- " + newMessage.data[0] + "- ";
  }
  return newMessage;
});
```

**Reply by megahertz** [reply]
Data is still the same array, so should be cloned too

**Reply by tripower** [reply]
so, this my 2 ways which will work

```
let newMessage = Object.assign({}, message);
newMessage.data = JSON.parse(JSON.stringify(message.data));
```

```
const lodashClonedeep = require('lodash.clonedeep');
let newMessage = lodashClonedeep(message);
```

**Comment by itsdouges** [comment]
Hello - is there any way to get the path of the log folder? I don't see how it's possible to build the variables fn to generate the path like it does internally:

```
log.transports.file.resolvePathFn(variables.getPathVariables(process.platform))
```

Any help appreciated. Example code would be perfect!

I'm able to get it working by hacking internals but I don't like it:

```
import log from "electron-log";
import variables from "electron-log/src/main/transports/file/variables";

const pathVariables = variables.getPathVariables(process.platform);

console.log(
  log.transports.file.resolvePathFn({
    ...pathVariables,
    fileName: log.transports.file.fileName,
  })
);
```

**Reply by megahertz** [reply]
Hi, you can use `log.transports.file.getFile().path` for that

**Reply by itsdouges** [reply]
Okay interestingly `getFile()` blows up during module initialization (when I was logging it immediately for testing) but when I use it after it works fine. I'll take it! Thank you.

**Comment by panyann** [comment]
Hello,
I have some feedback! I'm using this solution in my application and I think it lacks ability to keep only X number of lines in the file.
At the moment it seems like the file will grow forever, making my application bigger and bigger.
Or am I wrong and there is some sort of protection? I'm worried about that.

**Reply by megahertz** [reply]
Hi. Unfortunately, you can set only max log size, not a line number limit. It won't be implemented since it may impact performance.

**Reply by panyann** [reply]
That sounds good enough, how to set that max size? Is there any by default? What happens when it reaches it, does it clean all file?
Thanks in advance 🙂

**Reply by megahertz** [reply]
https://github.com/megahertz/electron-log/blob/master/docs/transports/file.md

**Comment by shiyufeng2** [comment]
Hi,I can't initialize elecfreon-log.The following is the error message.
Error occurred in handler for '__ELECTRON_LOG__': No handler registered for '__ELECTRON_LOG__'  
electron-log isn't initialized in the main process. Please call log.initialize() before. Error invoking remote method '__ELECTRON_LOG__': No handler registered for '__ELECTRON_LOG__'

**Reply by megahertz** [reply]
Have you followed this instruction?

**Reply by shiyufeng2** [reply]
I have installed v5.0.0
and in main.js
import logfrom 'electron-log'
log.initialize();

**Reply by megahertz** [reply]
Make sure `log.initialize();` is  actually called before a BrowserWindow is created
